azurerm/internal/services/storage/storage_management_policy_resource.go

```diff
@@ -243,172 +311,334 @@ func resourceStorageManagementPolicyDelete(d *schema.ResourceData, meta interfac
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+
+	return &storage.ManagementPolicyFilter{
+		PrefixMatch:    utils.ExpandStringSlice(input["prefix_match"].(*schema.Set).List()),
+		BlobTypes:      utils.ExpandStringSlice(input["blob_types"].(*schema.Set).List()),
+		BlobIndexMatch: expandAzureRmStorageBlobIndexMatch(input["blob_index_match_tag"].(*schema.Set).List()),
+	}
+}
+
+func expandStorageManagementPolicyActions(inputs []interface{}) *storage.ManagementPolicyAction {
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+
+	return &storage.ManagementPolicyAction{
+		BaseBlob: expandStorageManagementPolicyActionsBaseBlob(input["base_blob"].([]interface{})),
+		Snapshot: expandStorageManagementPolicyActionsSnapshot(input["snapshot"].([]interface{})),
+		Version:  expandStorageManagementPolicyActionsVersion(input["version"].([]interface{})),
+	}
+}
+
+func expandStorageManagementPolicyActionsBaseBlob(inputs []interface{}) *storage.ManagementPolicyBaseBlob {
+	if len(inputs) == 0 {
+		return nil
+	}
+	input := inputs[0].(map[string]interface{})
+	result := storage.ManagementPolicyBaseBlob{}
+
+	if v, ok := input["delete_after_days_since_modification_greater_than"].(float64); ok && v != -1 {
+		result.Delete = &storage.DateAfterModification{
+			DaysAfterModificationGreaterThan: utils.Float(v),
 		}
 	}
-	if _, ok := d.GetOk(fmt.Sprintf("rule.%d.actions", ruleIndex)); ok {
-		if _, ok := d.GetOk(fmt.Sprintf("rule.%d.actions.0.base_blob", ruleIndex)); ok {
-			baseBlob := &storage.ManagementPolicyBaseBlob{}
-			if v, ok := d.GetOk(fmt.Sprintf("rule.%d.actions.0.base_blob.0.tier_to_cool_after_days_since_modification_greater_than", ruleIndex)); ok {
-				if v != nil {
-					baseBlob.TierToCool = &storage.DateAfterModification{
-						DaysAfterModificationGreaterThan: utils.Float(float64(v.(int))),
-					}
-				}
-			}
-			if v, ok := d.GetOk(fmt.Sprintf("rule.%d.actions.0.base_blob.0.tier_to_archive_after_days_since_modification_greater_than", ruleIndex)); ok {
-				if v != nil {
-					baseBlob.TierToArchive = &storage.DateAfterModification{
-						DaysAfterModificationGreaterThan: utils.Float(float64(v.(int))),
-					}
-				}
-			}
-			if v, ok := d.GetOk(fmt.Sprintf("rule.%d.actions.0.base_blob.0.delete_after_days_since_modification_greater_than", ruleIndex)); ok {
-				if v != nil {
-					baseBlob.Delete = &storage.DateAfterModification{
-						DaysAfterModificationGreaterThan: utils.Float(float64(v.(int))),
-					}
-				}
-			}
-			definition.Actions.BaseBlob = baseBlob
+	if v, ok := input["tier_to_archive_after_days_since_modification_greater_than"].(float64); ok && v != -1 {
+		result.TierToArchive = &storage.DateAfterModification{
+			DaysAfterModificationGreaterThan: utils.Float(v),
+		}
+	}
+	if v, ok := input["tier_to_cool_after_days_since_modification_greater_than"].(float64); ok && v != -1 {
```

Comment:
this means this value can't be removed, is that intentional?

Reply:
I suppose this means that this value could be removed when unset. Because by default, it'll set a `0` value to all the parameters. `0` is meaningful value for this service. I want to set `null` to this parameters when unset. Some variable will return error message `features not supported` when setting a default `0` value.